### PR TITLE
feat: gateway streaming handler

### PR DESCRIPTION
**Pull request by scott-dn**
#198 

#### DEMO
```bash
$ curl http://0.0.0.0:4069/bafybeifyjj2bjhtxmp235vlfeeiy7sz6rzyx3lervfk3ap2nyn4rggqgei -v > origin
$ curl https://0.0.0.0:80/bafybeifyjj2bjhtxmp235vlfeeiy7sz6rzyx3lervfk3ap2nyn4rggqgei -k -v > gateway

$ openssl dgst -sha256 origin | awk '/.+$/{print "0x"$2}'
0x38b4a497f9ceea9d4fd024ee0d16b9989b71be2d4e8e10eaa67091d5db32f33e

$ openssl dgst -sha256 gateway | awk '/.+$/{print "0x"$2}'
0x38b4a497f9ceea9d4fd024ee0d16b9989b71be2d4e8e10eaa67091d5db32f33e
```

### CODE REVIEW

File: crates/ursa-gateway/src/worker/cache.rs [hunk 1]
```diff
@@ -69,30 +93,80 @@ impl WorkerCache for Cache {
...
+                    body,
+                ) => body,
+                resp => {
+                    error!("Error requested provider: {resp:?}");
+                    bail!("Error requested provider");
+                }
+            };
+            let key = String::from(k); // move to [worker|writer] thread
+            let tx = self.tx.clone(); // move to [worker|writer] thread
+            spawn(async move {
```
Comment (kckeiks) [inline crates/ursa-gateway/src/worker/cache.rs:144]: Isn't this still buffering the response from the cache node in memory?

Comment (kckeiks) [reply crates/ursa-gateway/src/worker/cache.rs:144]: This is a streaming feature though so nvm. We can discuss further if we need piping.

**Review by kckeiks — APPROVED** [review]

File: crates/ursa-gateway/src/worker/cache.rs (hunk 1 above)
Comment (scott-dn) [reply crates/ursa-gateway/src/worker/cache.rs:144]: yup, it's a streaming feature; we might need a pipeline solution one day.

File: crates/ursa-gateway/src/worker/cache.rs [hunk 2]
```diff
@@ -69,30 +93,80 @@ impl WorkerCache for Cache {
...
+                            bytes.put(buf);
+                        }
+                        Err(e) => {
+                            error!("Failed to read stream for {e:?}");
+                            return;
+                        }
+                    }
+                }
+                if let Err(e) = tx.send(WorkerCacheCommand::InsertSync {
+                    key,
```
Comment (kckeiks) [inline crates/ursa-gateway/src/worker/cache.rs:162]: Does each chunk get a different key (cid)?

Comment (scott-dn) [reply crates/ursa-gateway/src/worker/cache.rs:162]: Yes, each will have diff cid.

File: crates/ursa-gateway/src/server/route/api/v1/get.rs [hunk 3]
```diff
@@ -1,28 +1,47 @@
...
 use cid::Cid;
-use hyper::StatusCode;
 use serde_json::{json, Value};
 use tokio::sync::RwLock;
 
-use super::super::super::super::model::HttpResponse;
-use crate::worker::cache::ServerCache;
+use crate::{server::model::HttpResponse, worker::cache::ServerCache};
 
 pub async fn get_block_handler<Cache: ServerCache>(
```
Comment (b0xtch) [inline crates/ursa-gateway/src/server/route/api/v1/get.rs:16]:
```suggestion
pub async fn get_car_handler<Cache: ServerCache>(
```

**Review by b0xtch — APPROVED** [review]
Mint, regarding the generic cache. Would it be better to create an impl for IPLD with support for`Tlrfu`. Food for thought only because this already support a generic `v` value. With deduplication, so pretty good. 

Example:

```rust
impl Tlrfu for IPLD {
   //
}
```

File: crates/ursa-gateway/src/server/route/api/v1/get.rs (hunk 3 above)
Comment (scott-dn) [reply crates/ursa-gateway/src/server/route/api/v1/get.rs:16]: Good catch 👍

**Review by kckeiks — APPROVED** [review]

**Pull request merged**